// @intellectronica/ruler/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.COPILOT_SUBAGENTS_PATH = exports.CODEX_SUBAGENTS_PATH = exports.CURSOR_SUBAGENTS_PATH = exports.CLAUDE_SUBAGENTS_PATH = exports.RULER_SUBAGENTS_PATH = exports.SKILL_MD_FILENAME = exports.ANTIGRAVITY_SKILLS_PATH = exports.FACTORY_SKILLS_PATH = exports.WINDSURF_SKILLS_PATH = exports.CURSOR_SKILLS_PATH = exports.JUNIE_SKILLS_PATH = exports.GEMINI_SKILLS_PATH = exports.ROO_SKILLS_PATH = exports.VIBE_SKILLS_PATH = exports.GOOSE_SKILLS_PATH = exports.PI_SKILLS_PATH = exports.OPENCODE_SKILLS_PATH = exports.CODEX_SKILLS_PATH = exports.CLAUDE_SKILLS_PATH = exports.RULER_SKILLS_PATH = exports.SKILLS_DIR = exports.DEFAULT_RULES_FILENAME = exports.ERROR_PREFIX = void 0;
exports.actionPrefix = actionPrefix;
exports.createRulerError = createRulerError;
exports.logVerbose = logVerbose;
exports.logInfo = logInfo;
exports.logWarn = logWarn;
exports.logError = logError;
exports.logVerboseInfo = logVerboseInfo;
exports.ERROR_PREFIX = '[ruler]';
// Centralized default rules filename. Now points to 'AGENTS.md'.
// Legacy '.ruler/instructions.md' is still supported as a fallback with a warning.
exports.DEFAULT_RULES_FILENAME = 'AGENTS.md';
function actionPrefix(dry) {
    return dry ? '[ruler:dry-run]' : '[ruler]';
}
function createRulerError(message, context) {
    const fullMessage = context
        ? `${exports.ERROR_PREFIX} ${message} (Context: ${context})`
        : `${exports.ERROR_PREFIX} ${message}`;
    return new Error(fullMessage);
}
function logVerbose(message, isVerbose) {
    if (isVerbose) {
        console.error(`[ruler:verbose] ${message}`);
    }
}
/**
 * Centralized logging functions with consistent output streams and prefixing.
 * - info/verbose go to stdout (user-visible progress)
 * - warn/error go to stderr (problems)
 */
function logInfo(message, dryRun = false) {
    const prefix = actionPrefix(dryRun);
    console.log(`${prefix} ${message}`);
}
function logWarn(message, dryRun = false) {
    const prefix = actionPrefix(dryRun);
    console.warn(`${prefix} ${message}`);
}
function logError(message, dryRun = false) {
    const prefix = actionPrefix(dryRun);
    console.error(`${prefix} ${message}`);
}
function logVerboseInfo(message, isVerbose, dryRun = false) {
    if (isVerbose) {
        const prefix = actionPrefix(dryRun);
        console.log(`${prefix} ${message}`);
    }
}
// Skills-related constants
exports.SKILLS_DIR = 'skills';
exports.RULER_SKILLS_PATH = '.ruler/skills';
exports.CLAUDE_SKILLS_PATH = '.claude/skills';
exports.CODEX_SKILLS_PATH = '.codex/skills';
exports.OPENCODE_SKILLS_PATH = '.opencode/skills';
exports.PI_SKILLS_PATH = '.pi/skills';
exports.GOOSE_SKILLS_PATH = '.agents/skills';
exports.VIBE_SKILLS_PATH = '.vibe/skills';
exports.ROO_SKILLS_PATH = '.roo/skills';
exports.GEMINI_SKILLS_PATH = '.gemini/skills';
exports.JUNIE_SKILLS_PATH = '.junie/skills';
exports.CURSOR_SKILLS_PATH = '.cursor/skills';
exports.WINDSURF_SKILLS_PATH = '.windsurf/skills';
exports.FACTORY_SKILLS_PATH = '.factory/skills';
exports.ANTIGRAVITY_SKILLS_PATH = '.agent/skills';
exports.SKILL_MD_FILENAME = 'SKILL.md';
// Subagents-related constants
exports.RULER_SUBAGENTS_PATH = '.ruler/agents';
exports.CLAUDE_SUBAGENTS_PATH = '.claude/agents';
exports.CURSOR_SUBAGENTS_PATH = '.cursor/agents';
exports.CODEX_SUBAGENTS_PATH = '.codex/agents';
exports.COPILOT_SUBAGENTS_PATH = '.github/agents';
