// @intellectronica/ruler/dist/core/SubagentsUtils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseFrontmatter = parseFrontmatter;
exports.validateFrontmatter = validateFrontmatter;
exports.loadSubagentFile = loadSubagentFile;
exports.mapToolsForCopilot = mapToolsForCopilot;
const path = __importStar(require("path"));
const fs = __importStar(require("fs/promises"));
const yaml = __importStar(require("js-yaml"));
const FRONTMATTER_RE = /^---\r?\n([\s\S]*?)\r?\n---\r?\n?([\s\S]*)$/;
/**
 * Extracts YAML frontmatter and body from a Markdown file's contents.
 * Returns null if no frontmatter delimiter pair is present at the head of the file.
 */
function parseFrontmatter(content) {
    const match = FRONTMATTER_RE.exec(content);
    if (!match) {
        return null;
    }
    const [, raw, body] = match;
    const meta = yaml.load(raw);
    return {
        meta: meta && typeof meta === 'object' ? meta : {},
        body: body ?? '',
    };
}
/**
 * Validates a parsed frontmatter object and reads the required and optional
 * fields into a typed SubagentFrontmatter. Returns the typed value on success
 * or an error message on failure.
 */
function validateFrontmatter(meta, expectedName) {
    const name = meta.name;
    const description = meta.description;
    if (typeof name !== 'string' || name.length === 0) {
        return { error: `missing or invalid required field "name"` };
    }
    if (typeof description !== 'string' || description.length === 0) {
        return { error: `missing or invalid required field "description"` };
    }
    if (name !== expectedName) {
        return {
            error: `frontmatter name "${name}" does not match filename stem "${expectedName}"`,
        };
    }
    const fm = { name, description };
    if (meta.tools !== undefined) {
        if (Array.isArray(meta.tools) &&
            meta.tools.every((t) => typeof t === 'string')) {
            fm.tools = meta.tools;
        }
        else if (typeof meta.tools === 'string') {
            fm.tools = meta.tools
                .split(',')
                .map((t) => t.trim())
                .filter(Boolean);
        }
        else {
            return { error: `invalid "tools" field; expected string or string[]` };
        }
    }
    if (meta.model !== undefined) {
        if (typeof meta.model !== 'string') {
            return { error: `invalid "model" field; expected string` };
        }
        fm.model = meta.model;
    }
    if (meta.readonly !== undefined) {
        if (typeof meta.readonly !== 'boolean') {
            return { error: `invalid "readonly" field; expected boolean` };
        }
        fm.readonly = meta.readonly;
    }
    if (meta.is_background !== undefined) {
        if (typeof meta.is_background !== 'boolean') {
            return { error: `invalid "is_background" field; expected boolean` };
        }
        fm.is_background = meta.is_background;
    }
    return { value: fm };
}
/**
 * Loads a single subagent file and produces a SubagentInfo.
 * Invalid files produce a SubagentInfo with valid=false and an error string.
 */
async function loadSubagentFile(filePath) {
    const stem = path.basename(filePath, '.md');
    const content = await fs.readFile(filePath, 'utf8');
    // js-yaml throws on malformed YAML; convert that into the standard
    // validation-failure shape so one bad file doesn't abort discovery.
    let parsed;
    try {
        parsed = parseFrontmatter(content);
    }
    catch (error) {
        const detail = error instanceof Error ? error.message : String(error);
        return {
            name: stem,
            path: filePath,
            valid: false,
            error: `${stem}.md: invalid YAML frontmatter: ${detail}`,
        };
    }
    if (!parsed) {
        return {
            name: stem,
            path: filePath,
            valid: false,
            error: `${stem}.md: missing YAML frontmatter`,
        };
    }
    const result = validateFrontmatter(parsed.meta, stem);
    if ('error' in result) {
        return {
            name: stem,
            path: filePath,
            valid: false,
            error: `${stem}.md: ${result.error}`,
        };
    }
    return {
        name: stem,
        path: filePath,
        valid: true,
        frontmatter: result.value,
        body: parsed.body,
    };
}
/**
 * Maps Claude Code tool names to GitHub Copilot tool aliases.
 * Unknown source tools return undefined and should be dropped (with a warning).
 */
const COPILOT_TOOL_MAP = {
    Read: 'read',
    Grep: 'search',
    Glob: 'search',
    Bash: 'execute',
    Edit: 'edit',
    Write: 'edit',
    WebFetch: 'web',
    WebSearch: 'web',
    TodoWrite: 'todo',
    Task: 'agent',
};
/**
 * Translates Claude tool names to Copilot aliases. Deduplicates results.
 * Unknown source tools are reported separately so callers can surface a warning.
 */
function mapToolsForCopilot(sourceTools) {
    const mapped = new Set();
    const unknown = [];
    for (const tool of sourceTools) {
        const alias = COPILOT_TOOL_MAP[tool];
        if (alias) {
            mapped.add(alias);
        }
        else {
            unknown.push(tool);
        }
    }
    return { tools: Array.from(mapped), unknown };
}
